mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -19,118 +19,551 @@
-    local cherrypicks_change_godeps="false"
-    while read commit; do
-        if [[ -z "${commit}" ]]; then
-            continue
+    local dst_branch="${3}"
+    local kubernetes_remote="${4:-https://github.com/kubernetes/kubernetes.git}"
+    local deps="${5:-""}"
+    local is_library="${6}"
+    readonly subdirectory src_branch dst_branch kubernetes_remote deps is_library
+
+    local new_branch="false"
+    if ! git rev-parse -q --verify HEAD; then
+        echo "Found repo without ${dst_branch} branch, creating initial commit."
+        git commit -m "Initial commit" --allow-empty
+        new_branch="true"
+    else
+        echo "Starting at existing ${dst_branch} commit $(git rev-parse HEAD)."
+    fi
+
+    local dst_old_head=$(git rev-parse HEAD) # will be the initial commit for new branch
+
+    # fetch upstream kube and checkout $src_branch, name it filtered-branch
+    git remote rm upstream-kube >/dev/null || true
+    git remote add upstream-kube "${kubernetes_remote}" >/dev/null
+    git fetch -q upstream-kube --no-tags
+    git branch -D filtered-branch >/dev/null || true
+    git branch -f upstream-branch upstream-kube/"${src_branch}"
+    echo "Checked out k8s.io/kubernetes commit $(git rev-parse upstream-branch)."
+    git checkout -q upstream-branch -b filtered-branch
+    git reset -q --hard upstream-branch
+
+    # filter filtered-branch (= ${src_branch}) by ${subdirectory} modifying commits
+    # and rewrite paths. Each filtered commit (which is not dropped), gets the
+    # original k8s.io/kubernetes commit hash in the commit message as "Kubernetes-commit: <hash>".
+    # Then select all new mainline commits on filtered-branch as ${f_mainline_commits}
+    # to loop through them later.
+    local f_mainline_commits=""
+    if [ "${new_branch}" = "true" ] && [ "${src_branch}" = master ]; then
+	# new master branch
+	filter-branch ${src_branch} "${subdirectory}" filtered-branch
+
+	# find commits on the main line (mostly merges, some commits due to fast-forward merges that filter-branch dropped)
+        f_mainline_commits=$(git log --first-parent --format='%H' --reverse HEAD)
+
+	# create and checkout new, empty master branch
+	git checkout -q ${dst_branch} --orphan
+    else
+	# create filtered-branch-base before filtering for
+	# - new branches that branch off master (i.e. the branching point)
+	# - old branch which continue with the last old commit.
+	if [ "${new_branch}" = "true" ]; then
+	    # new non-master branch
+            local k_branch_point_commit=$(git-fork-point upstream-kube/${src_branch} upstream-kube/master)
+            if [ -z "${k_branch_point_commit}" ]; then
+		echo "Couldn't find a branch point of upstream-kube/${src_branch} and upstream-kube/master."
+		return 1
+            fi
+            echo "Using branch point ${k_branch_point_commit} as new starting point for new branch ${dst_branch}."
+	    git branch -f filtered-branch-base ${k_branch_point_commit} >/dev/null
```

Comment:
nit: missing a space in the the indention for this and the following two lines.

Reply:
had some tabs in the code. Untabified now.